### ISSUE TITLE
Error while loading key by handle

### ISSUE THREAD

**Issue by acartapanis** [aws-samples/aws-cloudhsm-jce-examples#35]
After updating to 3.1.0, we have an error while loading a key by key handle.

The exception is :
```
java.lang.ClassCastException: [B cannot be cast to [Ljava.lang.Byte;
  at com.cavium.key.CaviumKeyAttributes.getModulus(CaviumKeyAttributes.java:138)
  at com.cavium.key.CaviumRSAPrivateKey.<init>(CaviumRSAPrivateKey.java:48)
  at xxxx.xxxxx.xxxxx.CaviumUtils.getKeyByHandle(CaviumUtils.java:241)
```

The CaviumUtils.getKeyByHandle is an exact copy of https://github.com/aws-samples/aws-cloudhsm-jce-examples/blob/master/src/main/java/com/amazonaws/cloudhsm/examples/KeyUtilitiesRunner.java#L230

After some investigation, we find something strange in the CaviumKeyAttributes.getModulus method (CaviumKeyAttributes.java:106), specially when we compare it with the CloudHsmKeyAttributesMap.calculateModulus method (CloudHsmKeyAttributesMap.java:118).

**Comment by bnjsmt** [comment]
Thank you for highlighting this issue, @acartapanis .

This issue has been successfully reproduced.  Permanent fixes are being drafted.

In the meanwhile, there is a workaround.  Simply apply these changes to `KeyUtilitiesRunner.java` and `CaviumUtils.java`:

```diff
diff --git src/main/java/com/amazonaws/cloudhsm/examples/KeyUtilitiesRunner.java src/main/java/com/amazonaws/cloudhsm/examples/KeyUtilitiesRunner.java
index ae3df32..9d8d2f7 100644
--- src/main/java/com/amazonaws/cloudhsm/examples/KeyUtilitiesRunner.java
+++ src/main/java/com/amazonaws/cloudhsm/examples/KeyUtilitiesRunner.java
@@ -16,6 +16,7 @@
  */
 package com.amazonaws.cloudhsm.examples;
 
+import com.amazonaws.cloudhsm.CloudHsmKeyAttributes;
 import com.cavium.cfm2.CFM2Exception;
 import com.cavium.cfm2.ImportKey;
 import com.cavium.cfm2.Util;
@@ -235,27 +236,27 @@ public class KeyUtilitiesRunner {
         CaviumKeyAttributes cka = new CaviumKeyAttributes(keyAttribute);
 
         if(cka.getKeyType() == CaviumKeyAttributes.KEY_TYPE_AES) {
-            CaviumAESKey aesKey = new CaviumAESKey(handle, cka);
+            CaviumAESKey aesKey = new CaviumAESKey(handle, CloudHsmKeyAttributes.fromCavium(cka.getBytes()));
             return aesKey;
         }
         else if(cka.getKeyType() == CaviumKeyAttributes.KEY_TYPE_RSA && cka.getKeyClass() == CaviumKeyAttributes.CLASS_PRIVATE_KEY) {
-            CaviumRSAPrivateKey privKey = new CaviumRSAPrivateKey(handle, cka);
+            CaviumRSAPrivateKey privKey = new CaviumRSAPrivateKey(handle, CloudHsmKeyAttributes.fromCavium(cka.getBytes()));
             return privKey;
         }
         else if(cka.getKeyType() == CaviumKeyAttributes.KEY_TYPE_RSA && cka.getKeyClass() == CaviumKeyAttributes.CLASS_PUBLIC_KEY) {
-            CaviumRSAPublicKey pubKey = new CaviumRSAPublicKey(handle, cka);
+            CaviumRSAPublicKey pubKey = new CaviumRSAPublicKey(handle, CloudHsmKeyAttributes.fromCavium(cka.getBytes()));
             return pubKey;
         }
         else if(cka.getKeyType() == CaviumKeyAttributes.KEY_TYPE_EC && cka.getKeyClass() == CaviumKeyAttributes.CLASS_PRIVATE_KEY) {
-            CaviumECPrivateKey privKey = new CaviumECPrivateKey(handle, cka);
+            CaviumECPrivateKey privKey = new CaviumECPrivateKey(handle, CloudHsmKeyAttributes.fromCavium(cka.getBytes()));
             return privKey;
         }
         else if(cka.getKeyType() == CaviumKeyAttributes.KEY_TYPE_EC && cka.getKeyClass() == CaviumKeyAttributes.CLASS_PUBLIC_KEY) {
-            CaviumECPublicKey pubKey = new CaviumECPublicKey(handle, cka);
+            CaviumECPublicKey pubKey = new CaviumECPublicKey(handle, CloudHsmKeyAttributes.fromCavium(cka.getBytes()));
             return pubKey;
         }
         else if(cka.getKeyType() == CaviumKeyAttributes.KEY_TYPE_GENERIC_SECRET) {
-            CaviumKey key = new CaviumAESKey(handle, cka);
+            CaviumKey key = new CaviumAESKey(handle, CloudHsmKeyAttributes.fromCavium(cka.getBytes()));
             return key;
         }
```

**Comment by acartapanis** [comment]
Hi, may I know why the previous comment was deleted ?


Are we supposed to open a case on aws support for this issue ?

**Comment by sveerg** [comment]
Facing the same issue after the upgrade, what was the work around that worked for you?

**Comment by gmudambi** [comment]
This issue has been addressed with software version 3.1.1.

You can find the links to package and release notes here:
https://docs.aws.amazon.com/cloudhsm/latest/userguide/client-history.html#client-version-3-1-1

Marking this issues as resolved.
